snuba/replacer.py

```diff
@@ -277,3 +280,63 @@ def process_unmerge(message):
     query_time_flags = (NEEDS_FINAL, message['project_id'])
 
     return (count_query_template, insert_query_template, query_args, query_time_flags)
+
+
+def process_delete_tag(message):
+    tag = message['tag']
+    if not tag:
+        return None
+
+    assert isinstance(tag, six.string_types)
+    timestamp = datetime.strptime(message['datetime'], settings.PAYLOAD_DATETIME_FORMAT)
+    is_promoted = tag in PROMOTED_TAGS['tags']
+
+    where = """\
+        WHERE project_id = %(project_id)s
+        AND received <= CAST('%(timestamp)s' AS DateTime)
+        AND NOT deleted
+    """
+
+    if is_promoted:
+        where += "AND %(tag_column)s IS NOT NULL"
+    else:
+        where += "AND has(`tags.key`, %(tag_str)s)"
+
+    insert_query_template = """\
+        INSERT INTO %(dist_table_name)s (%(all_columns)s)
+        SELECT %(select_columns)s
+        FROM %(dist_table_name)s FINAL
+    """ + where
+
+    select_columns = []
+    for col in ALL_COLUMNS:
+        if is_promoted and col.flattened == tag:
+            select_columns.append('NULL')
+        elif col.flattened == 'tags.key':
+            select_columns.append(
+                "arrayFilter(x -> (indexOf(`tags.key`, x) != indexOf(`tags.key`, %s)), `tags.key`)" % escape_string(tag)
+            )
+        elif col.flattened == 'tags.value':
+            select_columns.append(
+                "arrayFilter(x -> (indexOf(`tags.value`, x) != indexOf(`tags.key`, %s)), `tags.value`)" % escape_string(tag)
+            )
+        else:
+            select_columns.append(col.escaped)
+
+    query_args = {
+        'all_columns': ', '.join(ALL_COLUMN_NAMES),
+        'select_columns': ', '.join(select_columns),
+        'project_id': message['project_id'],
+        'tag_str': escape_string(tag),
+        'tag_column': escape_col(tag),
```

Comment:
I think we maybe need to do a little more translation here? Probably have to at least lookup the tag in `TAG_COLUMN_MAP` to get the clickhouse column name from the tag provided.

A test to make sure that stuff like `browser.name` / `browser_name` works would be good too.

Reply:
Good catch, fixed.